src/VecSim/algorithms/ivf/ivf.cuh

```diff
@@ -0,0 +1,288 @@
+                      inner.kmeans_trainset_fraction = ivfParams->kmeans_trainsetFraction;
+                      inner.conservative_memory_allocation =
+                          ivfParams->conservativeMemoryAllocation;
+                      if constexpr (std::is_same_v<decltype(inner),
+                                                   raft::neighbors::ivf_pq::index_params>) {
+                          inner.pq_bits = ivfParams->pqBits;
+                          inner.pq_dim = ivfParams->pqDim;
+                          inner.codebook_kind = GetRaftCodebookKind(ivfParams->codebookKind);
+                      } else {
+                          inner.adaptive_centers = ivfParams->adaptiveCenters;
+                      }
+                  },
+                  result);
+              return result;
+          }()},
+          search_params_{[ivfParams]() {
+              auto result = ivfParams->pqBits > 0 ? search_params_t{std::in_place_index<1>}
+                                                  : search_params_t{std::in_place_index<0>};
+              std::visit(
+                  [ivfParams](auto &&inner) {
+                      inner.n_probes = ivfParams->nProbes;
+                      if constexpr (std::is_same_v<decltype(inner),
+                                                   raft::neighbors::ivf_pq::search_params>) {
+                          inner.lut_dtype = GetCudaType(ivfParams->lutType);
+                          inner.internal_distance_dtype =
+                              GetCudaType(ivfParams->internalDistanceType);
+                          inner.preferred_shmem_carvout = ivfParams->preferredShmemCarveout;
+                      }
+                  },
+                  result);
+              return result;
+          }()},
+          index_{std::nullopt} {}
+    auto addVector(const void *vector_data, labelType label,
+                   bool overwrite_allowed = true) override {
+        return addVectorBatch(vector_data, &label, 1, overwrite_allowed);
+    }
+    auto addVectorBatchAsync(const void *vector_data, labelType *label, size_t batch_size,
+                             bool overwrite_allowed = true) {
+        // Allocate memory on device to hold vectors to be added
+        auto vector_data_gpu =
+            raft::make_device_matrix<data_type, internal_idx_t>(res_, batch_size, this->dim);
+        // Allocate memory on device to hold vector labels
+        auto label_gpu = raft::make_device_vector<labelType>(res_, batch_size);
+
+        // Copy vector data to previously allocated device buffer
+        raft::copy(vector_data_gpu.data_handle(), static_cast<DataType const *>(vector_data),
+                   this->dim * batch_size, res_.get_stream());
+        // Copy label data to previously allocated device buffer
+        raft::copy(label_gpu.data_handle(), label, batch_size, res_.get_stream());
+
+        if (std::holds_alternative<raft::neighbors::ivf_flat::index_params>(build_params_)) {
```

Comment:
Is there a way to avoid this `if`?

Reply:
Not unless we want to split IVF flat and IVF PQ back out into separate indexes again.

One possible alternate spelling would be to use `std::visit` with a constexpr if to distinguish the types. This has the disadvantage that compilers sometimes have trouble creating an efficient `std::visit` over small variants, but it has the advantage of providing obviously idiomatic access to the variant.

If performance is what we're worried about, I would avoid `std::visit` but add a compile-time assumption (via `__builtin_unreachable`, `__assume` or similar based on the compiler) that the variant contains the expected type prior to the `std::get` call. That usually allows the compiler to avoid the check that it must ordinarily make to see if it needs to throw a `std::bad_variant_access`.

If readability is what we are concerned about, the bodies of those if clauses could be separated into their own functions.

Reply:
Can you please share a suggestion for the last proposal, so we can iterate? (could be here or on slack).  I'm thinking about modularity and code reusability here.